src/cpp/src/threaded_streamer.hpp

```diff
@@ -0,0 +1,138 @@
+// Copyright (C) 2025 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <condition_variable>
+#include <queue>
+#include <thread>
+
+#include "openvino/genai/llm_pipeline.hpp"
+#include "openvino/genai/tokenizer.hpp"
+#include "text_callback_streamer.hpp"
+
+namespace ov {
+namespace genai {
+
+class ThreadedStreamerWrapper {
+public:
+    ThreadedStreamerWrapper(const StreamerVariant& streamer, Tokenizer& tokenizer) {
+        if (auto streamer_obj = std::get_if<std::monostate>(&streamer)) {
+            m_streamer_ptr = nullptr;
+        } else if (auto streamer_obj = std::get_if<std::shared_ptr<StreamerBase>>(&streamer)) {
+            m_streamer_ptr = *streamer_obj;
+        } else if (auto callback = std::get_if<std::function<bool(std::string)>>(&streamer)) {
+            m_streamer_ptr = std::make_shared<TextCallbackStreamer>(tokenizer, *callback);
+        }
+    }
+
+    void start() {
+        if (!m_streamer_ptr) {
+            return;
+        }
+
+        m_worker_thread = std::make_shared<std::thread>(&ThreadedStreamerWrapper::_worker, this);
+    }
+
+    void put(const std::vector<int64_t>& tokens) {
+        if (!m_streamer_ptr) {
+            return;
+        }
+
+        std::lock_guard<std::mutex> lock(m_mutex);
+        m_queue.push(tokens);
+        m_cv.notify_one();
+    }
+
+    void put(const int64_t token) {
+        if (!m_streamer_ptr) {
+            return;
+        }
+
+        std::lock_guard<std::mutex> lock(m_mutex);
+        m_queue.push(token);
+        m_cv.notify_one();
+    }
+
+    void end() {
+        if (!m_streamer_ptr) {
+            return;
+        }
+
+        {
+            std::lock_guard<std::mutex> lock(m_mutex);
+            m_stopped = true;
+        }
+
+        m_cv.notify_one();
+
+        if (m_worker_thread && m_worker_thread->joinable()) {
+            m_worker_thread->join();
+        }
+
+        m_streamer_ptr->end();
+    }
+
+    bool is_dropped() {
+        std::lock_guard<std::mutex> lock(m_mutex);
+        return m_dropped;
+    }
+
+    bool has_callback() {
+        return static_cast<bool>(m_streamer_ptr);
+    }
+
+private:
+    std::shared_ptr<StreamerBase> m_streamer_ptr = nullptr;
+    std::shared_ptr<std::thread> m_worker_thread = nullptr;
+    std::mutex m_mutex;
+    std::condition_variable m_cv;
+    std::queue<std::variant<int64_t, std::vector<int64_t>>> m_queue;
```

Comment:
Why not `SynchronizedQueue` from `openvino.genai/src/cpp/src/synchronized_queue.hpp`?

Reply:
There was a deadlock for `squeue.pull`. It happen on `streamer.end` call.
1. streamer thread waits for a new token
2. main thread calls `streamer.end`, streamer thread should be stopped. But there is no API to gracefully unlock squeue.

One way to unlock squeue is to push some dummy token and handle `is_stopped` flag.
I thought it may be not very clean.

I guess the trick with dummy token and sync queue can simplify implementation. Do you want me to implement that?

Reply:
Look like we handled it via https://github.com/openvinotoolkit/openvino.genai/blob/a0852d0d006daaf55412a877d0cc70fd51a55770/src/cpp/src/sequence_group.hpp#L655

Reply:
Reimplemented with `SynchronizedQueue`